### PR TITLE
Add custom mapping error handler feature.

### PR DESCRIPTION
**Pull request by adjam**
Adds a new setting that takes a lambda which is invoked when an unhandled exception occurs in `map_record`.

### CODE REVIEW

**Comment by jrochkind** [comment]
You can actually get the index_step from context.index_step, at least you should be able to unless it's broken, so doesn't need to be passed separately. Hint docs might be good though.

File: README.md [hunk 1]
```diff
@@ -363,6 +363,35 @@ Fields that are empty will have a value sent to the writer of an empty
...
+exception, e.g.
+
+``` 
+# log errors until we exceed some value
+max_error_count = 5
+errors = 0
+  indexer.settings['mapping_error_handler] = lambda do |ctx, step, e|
+    errors += 1
+      ctx.logger.warn("I got #{e} during #{step.inspect} but that is not going to bother me one bit")
+      if errors > max_error_count
```
Comment (jrochkind) [inline README.md]: believe this is supposed to be `<` than `>`?

Comment (adjam) [reply README.md]: Yeah, probably, if you're into the whole "do what the variable name suggests" thing =)

Comment (adjam) [reply README.md]: The signature of the lambda matches what was in scope for the original `log_mapping_errors` method, with the exception tacked onto the end: https://github.com/traject/traject/blob/master/lib/traject/indexer.rb#L365

Comment (jrochkind) [reply README.md]: It's possible the original didn't really need that second param either. Note that the code that originally calls `log_mapping_errors` sets `context.index_step` to the same thing as second param, right before it calls it and passes context in too. 

https://github.com/traject/traject/blob/d7d010747a9fde23db77a2e7eaa7e321d0059405/lib/traject/indexer.rb#L344-L345

**Comment by jrochkind** [comment]
Have to think about this interacts with a different `rescue` I put into the `process_with` in 3.0. I think it makes the additional rescue API I added unneccesary, can just use this. But have to think more about how people can use the programmatic API with it being clear to the developer what the rescue behavior will be. 

I'm inclined to merge this to 3.0, rather than 2.0, so I don't need to think through two different situations and merges, and so I can use ruby keyword args in some places I'm thinking of using them to make it more clear under programmatic use what the rescue behavior is, and easily override it.

**Comment by jrochkind** [comment]
Oh, also that max_records example actually ought to be done with a thread-safe counter, in case there are threads going on. https://github.com/traject/traject/blob/d7d010747a9fde23db77a2e7eaa7e321d0059405/lib/traject/solr_json_writer.rb#L153

**Comment by jrochkind** [comment]
I want to port this over to the 3.0 branch, and am getting lost trying to rebase. Sorry, I'll see if I can get your name in the commit anyway!

**Comment by jrochkind** [comment]
Closed in favor of #178 for traject 3.0. Thanks @adjam !

**Pull request closed without merging**